README.md

```diff
@@ -5,26 +5,77 @@
 [![License](https://img.shields.io/cocoapods/l/swift-async-queue.svg)](https://cocoapods.org/pods/swift-async-queue)
 [![Platform](https://img.shields.io/cocoapods/p/swift-async-queue.svg)](https://cocoapods.org/pods/swift-async-queue)
 
-A queue that enables sending FIFO-ordered tasks from synchronous to asynchronous contexts.
+A library of queues that enables sending ordered tasks from synchronous to asynchronous contexts.
 
-## Usage
+## Task Ordering and Swift Concurrency
 
-### Basic Initialization
+Tasks sent from a synchronous context to an asynchronous context in Swift Concurrency are inherently unordered. Consider the following test:
 
-```swift
-let asyncQueue = AsyncQueue()
 ```
+@MainActor
+func test_mainActor_taskOrdering() async {
+    var counter = 0
+    var tasks = [Task<Void, Never>]()
+    for iteration in 1...100 {
+        tasks.append(Task {
+            counter += 1
+            XCTAssertEqual(counter, iteration) // often fails
+        })
+    }
+    for task in tasks {
+        _ = await task.value
+    }
+}
+```
+
+Despite the spawned `Task` inheriting the serial `@MainActor` execution context, the ordering of the scheduled asynchronous work is not guaranteed.
 
-### Sending events from a synchronous context
+While [actors](https://docs.swift.org/swift-book/LanguageGuide/Concurrency.html#ID645) are great at serializing tasks, there is no simple way in the standard Swift library to send ordered tasks to them from a synchronous context.
+
+### Executing asynchronous tasks in FIFO order
+
+Use a `FIFOQueue` queue to execute asynchronous tasks enqueued from a nonisolated context in FIFO order. Tasks sent to one of these queues are guaranteed to begin _and end_ executing in the order in which they are enqueued.
 
 ```swift
-asyncQueue.async { /* awaitable context that executes after all other enqueued work is completed */ }
+let queue = FIFOQueue()
+queue.async {
+    /*
+    `async` context that executes after all other enqueued work is completed.
+    Work enqueued after this task will wait for this task to complete.
+    */
+}
+Task {
+    await queue.await {
+        /*
+        `async` context that can return a value or throw an error.
+        Executes after all other enqueued work is completed.
+        Work enqueued after this task will wait for this task to complete.
+        */
+    }
+}
 ```
 
-### Awaiting work from an asynchronous context
+### Sending ordered asynchronous tasks to Actors
+
+Use an `ActorQueue` queue to send ordered asynchronous tasks from a nonisolated context to an `actor` instance. Tasks sent to one of these queues are guaranteed to begin _but not end_ executing in the order in which they are enqueued.
```

Comment:
I think going a bit more into detail would be good.
```suggestion
Use an `ActorQueue` queue to send ordered asynchronous tasks from a nonisolated context to an `actor` instance. Tasks sent to one of these queues are guaranteed to begin executing by the actor in the order in which they are enqueued. Whether those tasks will also end in the same order depends on the actor's implementation, as it may be using concurrency internally itself.
```

Reply:
See later comments though. This is a gnarly thing to define properly.

Reply:
We're basically talking about FIFE (first-in-first-executing) contract, with ordering guarantees up until the first suspension point. So maybe we define this as:
```suggestion
Use an `ActorQueue` queue to send ordered asynchronous tasks from a nonisolated context to an `actor` instance. Tasks sent to one of these queues are guaranteed to begin executing in the order in which they are enqueued. Ordering of execution is guaranteed up until the first [suspension point](https://docs.swift.org/swift-book/LanguageGuide/Concurrency.html#ID639) within the called `actor` code.
```

Reply:
Ok that is probably better, with concurrent logic around we always get to the point where we just 🤷🏽‍♂️ and hope what the caller is doing very soon.

I'd like to experiment with sending synchronous code to run, but I don't think Swift will let us as currently configured (maybe the actor can run a serial block block itself but how to declare that in a way that will always work I'm not sure).

Reply:
> I'd like to experiment with sending synchronous code to run, but I don't think Swift will let us as currently configured

This is doable! I've written + deleted this test a few times in the process of discovering + trying to delete the warning I put on this type. Looks like this:
```
func test_async_sendsSynchronousEventsInOrder() async {
    final class SynchronousCounter: @unchecked Sendable {
        func incrementAndExpectCount(equals expectedCount: Int) {
            increment()
            XCTAssertEqual(expectedCount, count)
        }

        func increment() {
            count += 1
        }

        var count: Int {
            set {
                dispatchQueue.async { self.unsafeCount = newValue }
            }
            get {
                dispatchQueue.sync { unsafeCount }
            }
        }
        var unsafeCount = 0
        let dispatchQueue = DispatchQueue(label: #function)
    }
    let counter = SynchronousCounter()
    for iteration in 1...1_000 {
        systemUnderTest.async {
            counter.incrementAndExpectCount(equals: iteration)
        }
    }
    await systemUnderTest.await { /* Drain the queue */ }
}
```

Reply:
Note that in the above code we don't really need the `dispatchQueue` barrier – I added it only to prevent a crash in case code were unexpectedly executed concurrently.